proposals/p2360.md

```diff
@@ -0,0 +1,312 @@
+Given a generic function with a generic type parameter, uses of that parameter
+will be implicitly converted to type `Type`. For example:
+
+```
+fn F[T:! Printable](x: T) { x.Print(); }
+```
+
+... is equivalent to ...
+
+```
+fn F[T:! Printable](x: T as Type) { x.Print(); }
+```
+
+As a result, we can no longer describe lookup for `x` as looking into the
+type-of-type, that is, the type of the expression after `x:`, because after
+implicit conversion, that expression is of type `Type`. Instead, lookup will
+look into the type of `x`, which symbolically evaluates to the value
+corresponding to `T` in type `Type`, and will look at that value to determine
+where else to look. Because that value is symbolically the name of a generic
+type parameter, we will look in the type of `T`, which is `Printable`.
+
+### Things not being changed
+
+This proposal does not introduce the name `Type`, which remains provisional. See
+[issue #2113](https://github.com/carbon-language/carbon-lang/issues/2113) for
+discussion of how things like `Type` should be named. At this time, that issue
+is still open, but it might conclude that `Type` is renamed `Core.Type`, with a
+keyword alias `type`.
+
+Under the [current design](/docs/design/generics/details.md#named-constraints),
+`Type` is the kind with no requirements, and an equivalent kind can be declared
+with `constraint MyVersionOfType {}`. This proposal does not change this
+decision. Because such a named constraint `MyVersionOfType` is equivalent to
+`Type`, that is, because they are the same kind, a value `T:! MyVersionOfType`
+is a type under this proposal.
+
+## Rationale
+
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   Making the behavior of tuples and empty structs more like other types is
+        easier for tooling to handle.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
+    -   The transition between templates and generics is made smoother by
+        changing compound member access to behave consistently regardless of
+        whether the left-hand operand is a type.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Bottoming out at the same type, `Type`, for all types is likely to be
+        easier to understand. However, having `(i32, i32)` not be the type of a
+        pair of integers, but instead be _implicitly convertible to_ the type of
+        a pair of integers, is likely to be harder to understand. It's unclear
+        which of these will be more dominant, but the new behavior is more
+        consistent.
+
+## Alternatives considered
+
+### Alternative terminology
+
+We considered various alternatives for the terminology choices listed above in
+[open discussion](https://docs.google.com/document/d/1tEt4iM6vfcY0O0DG0uOEMIbaXcZXlNREc2ChNiEtn_w/edit#heading=h.lecxe8qywdxa).
+
+Instead of "kind" we considered:
```

Comment:
Is this consistent with the existing [type-theoretic meaning](https://en.wikipedia.org/wiki/Kind_(type_theory)) of "kind"? It doesn't look like it to me, but I'm not very fluent with that part of type theory.

If not, it seems risky to repurpose a subtle type-theoretic term to describe a _different_ subtle corner of Carbon's type system,

Reply:
Yeah, I don't think this is entirely consistent with type-theoretical usage. What we're doing in Carbon at the kind level is very similar to the GHC [type-in-type](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/poly_kinds.html#overview-of-type-in-type) extension, which is basically saying that the type level and the kind level are the same thing. Suggestions for a better term would be appreciated!

Reply:
If this does not match the broader usage of the word "kind", that would be a big concern for me. (I have no confidence in my understanding of the broader usage myself though.)

Reply:
I don't have any ideas, but I wonder if it would help to have a term for "a type whose values can be implicitly converted to `Type`"? That seems like it might be a handy term in its own right, and might be a useful starting point for naming kinds (which are a subset of those things, IIUC).